File: server/routes/bucket.ts
import { Router, Response } from 'express';
import {
  GetBucketVersioningCommand,
  GetBucketEncryptionCommand,
  GetBucketLifecycleConfigurationCommand,
} from '@aws-sdk/client-s3';
import { s3Middleware, requireBucket, AuthenticatedRequest } from '../middleware/auth.js';

const router = Router();

function isEncryptionNotSupported(error: unknown): boolean {
  const err = error as {
    name?: string;
    code?: string;
    Code?: string;
    message?: string;
    $metadata?: { httpStatusCode?: number };
  };
  const httpStatus = err.$metadata?.httpStatusCode;
  if (httpStatus === 501) {
    return true;
  }
  const name = err.name ?? err.code ?? err.Code;
  if (name && ['NotImplemented', 'NotImplementedException', 'NotImplementedError', 'UnsupportedOperation'].includes(name)) {
    return true;
  }
  const message = typeof err.message === 'string' ? err.message.toLowerCase() : '';
  return message.includes('not implemented') || message.includes('notimplemented') || message.includes('unimplemented');
}

function isLifecycleNotConfigured(error: unknown): boolean {
  const err = error as {
    name?: string;
    code?: string;
    Code?: string;
    message?: string;
    $metadata?: { httpStatusCode?: number };
  };
  const httpStatus = err.$metadata?.httpStatusCode;
  if (httpStatus === 404) {
    return true;
  }
  const name = err.name ?? err.code ?? err.Code;
  if (name && ['NoSuchLifecycleConfiguration', 'NotFound'].includes(name)) {
    return true;
  }
  const message = typeof err.message === 'string' ? err.message.toLowerCase() : '';
  return message.includes('nosuchlifecycleconfiguration') || message.includes('no such lifecycle configuration');
}

interface LifecycleRule {
  id?: string;
  status: 'Enabled' | 'Disabled' | 'Unknown';
  prefix?: string;
  expiration?: {
    days?: number;
    date?: string;
    expiredObjectDeleteMarker?: boolean;
  };
  transitions?: Array<{
    days?: number;
    date?: string;
    storageClass: string;
  }>;
  noncurrentVersionExpiration?: {
    days?: number;
    newerNoncurrentVersions?: number;
  };
  abortIncompleteMultipartUpload?: {
    daysAfterInitiation?: number;
  };
}

interface BucketInfo {
  versioning: {
    status?: string;
    mfaDelete?: string;
  } | null;
  encryption: {
    algorithm?: string;
    kmsKeyId?: string;
  } | null;
  encryptionError?: string;
  lifecycleRules: LifecycleRule[];
}

// GET /api/bucket/:connectionId/:bucket/info
router.get('/:connectionId/:bucket/info', s3Middleware, requireBucket, async (req: AuthenticatedRequest, res: Response): Promise<void> => {
  const bucket = req.s3Credentials?.bucket;
  const client = req.s3Client;

  if (!bucket || !client) {
    res.status(500).json({ error: 'Internal server error' });
    return;
  }

  const result: BucketInfo = {
    versioning: null,
    encryption: null,
    lifecycleRules: [],
  };

  // Get versioning status
  try {
    const versioningCommand = new GetBucketVersioningCommand({ Bucket: bucket });
    const versioningResponse = await client.send(versioningCommand);
    result.versioning = {
      status: versioningResponse.Status,
      mfaDelete: versioningResponse.MFADelete,
    };
  } catch (err) {
    // If we can't get versioning, leave as null
    console.error('Failed to get bucket versioning:', err);
  }

  // Get encryption settings
  try {
    const encryptionCommand = new GetBucketEncryptionCommand({ Bucket: bucket });
    const encryptionResponse = await client.send(encryptionCommand);
    const rules = encryptionResponse.ServerSideEncryptionConfiguration?.Rules;
    if (rules && rules.length > 0) {
      const defaultRule = rules[0].ApplyServerSideEncryptionByDefault;
      if (defaultRule) {
        result.encryption = {
          algorithm: defaultRule.SSEAlgorithm,
          kmsKeyId: defaultRule.KMSMasterKeyID,
        };
      }
    }
  } catch (err: unknown) {
    const errorName = (err as { name?: string })?.name;
    if (errorName === 'ServerSideEncryptionConfigurationNotFoundError') {
      // No encryption configured - leave as null
    } else if (isEncryptionNotSupported(err)) {
      result.encryptionError = 'Not supported by this storage provider';
    } else {
      // Real error - store for display
      console.error('Failed to get bucket encryption:', err);
      const errorMessage = err instanceof Error ? err.message : String(err);
      result.encryptionError = errorName ? `${errorName}: ${errorMessage}` : errorMessage;
    }
  }

  // Get lifecycle rules
  try {
    const lifecycleCommand = new GetBucketLifecycleConfigurationCommand({ Bucket: bucket });
    const lifecycleResponse = await client.send(lifecycleCommand);
    if (lifecycleResponse.Rules) {
      result.lifecycleRules = lifecycleResponse.Rules.map((rule) => ({
        id: rule.ID,
        status: rule.Status ?? 'Unknown',
        prefix: rule.Filter?.Prefix,
        expiration: rule.Expiration ? {
          days: rule.Expiration.Days,
          date: rule.Expiration.Date?.toISOString(),
          expiredObjectDeleteMarker: rule.Expiration.ExpiredObjectDeleteMarker,
        } : undefined,
        transitions: rule.Transitions?.map((t) => ({
          days: t.Days,
          date: t.Date?.toISOString(),
          storageClass: t.StorageClass ?? 'Unknown',
        })),
        noncurrentVersionExpiration: rule.NoncurrentVersionExpiration ? {
          days: rule.NoncurrentVersionExpiration.NoncurrentDays,
          newerNoncurrentVersions: rule.NoncurrentVersionExpiration.NewerNoncurrentVersions,
        } : undefined,
        abortIncompleteMultipartUpload: rule.AbortIncompleteMultipartUpload ? {
          daysAfterInitiation: rule.AbortIncompleteMultipartUpload.DaysAfterInitiation,
        } : undefined,
      }));
    }
  } catch (err: unknown) {
    // Treat "not configured" as empty lifecycle rules (no log).
    if (!isLifecycleNotConfigured(err)) {
      console.error('Failed to get bucket lifecycle:', err);
    }
  }

  res.json(result);
});

export default router;

File: s3browser/routers/bucket.py
from fastapi import APIRouter, Depends

from s3browser.dependencies import get_s3_context
from s3browser.s3 import S3Context, error_code, error_message, http_status_code, require_bucket
from s3browser.utils import isoformat_z

router = APIRouter(prefix="/api/bucket", tags=["bucket"])

_ENCRYPTION_NOT_SUPPORTED_CODES = {
    "NotImplemented",
    "NotImplementedException",
    "NotImplementedError",
    "UnsupportedOperation",
}

_LIFECYCLE_NOT_CONFIGURED_CODES = {"NoSuchLifecycleConfiguration", "NotFound"}


def _is_encryption_not_supported(error: object) -> bool:
    if http_status_code(error) == 501:
        return True
    if error_code(error) in _ENCRYPTION_NOT_SUPPORTED_CODES:
        return True
    message = error_message(error).lower()
    return "not implemented" in message or "notimplemented" in message or "unimplemented" in message


def _is_lifecycle_not_configured(error: object) -> bool:
    status = http_status_code(error)
    code = error_code(error)
    if status == 404 and code in _LIFECYCLE_NOT_CONFIGURED_CODES:
        return True
    message = error_message(error).lower()
    if status == 404 and not code and "lifecycle" in message:
        return True
    return "nosuchlifecycleconfiguration" in message or "no such lifecycle configuration" in message


@router.get("/{connection_id}/{bucket}/info")
async def bucket_info(context: S3Context = Depends(get_s3_context)) -> dict[str, object]:
    bucket = require_bucket(context)
    client = context.client
    result: dict[str, object] = {
        "versioning": None,
        "encryption": None,
        "encryptionError": None,
        "lifecycleError": None,
        "lifecycleRules": [],
    }
    try:
        versioning = await client.get_bucket_versioning(Bucket=bucket)
        result["versioning"] = {
            "status": versioning.get("Status"),
            "mfaDelete": versioning.get("MFADelete"),
        }
    except Exception as error:
        print(f"Failed to get bucket versioning: {error}")
    try:
        encryption = await client.get_bucket_encryption(Bucket=bucket)
        rules = encryption.get("ServerSideEncryptionConfiguration", {}).get("Rules", [])
        if rules:
            default_rule = rules[0].get("ApplyServerSideEncryptionByDefault", {})
            if default_rule:
                result["encryption"] = {
                    "algorithm": default_rule.get("SSEAlgorithm"),
                    "kmsKeyId": default_rule.get("KMSMasterKeyID"),
                }
    except Exception as error:
        text = str(error)
        if "ServerSideEncryptionConfigurationNotFoundError" in text:
            pass
        elif _is_encryption_not_supported(error):
            result["encryptionError"] = "Not supported by this storage provider"
        else:
            print(f"Failed to get bucket encryption: {error}")
            result["encryptionError"] = text
    try:
        lifecycle = await client.get_bucket_lifecycle_configuration(Bucket=bucket)
        rules_out: list[dict[str, object]] = []
        for rule in lifecycle.get("Rules", []):
            expiration = rule.get("Expiration")
            transitions = rule.get("Transitions")
            noncurrent = rule.get("NoncurrentVersionExpiration")
            abort_upload = rule.get("AbortIncompleteMultipartUpload")
            item: dict[str, object] = {
                "id": rule.get("ID"),
                "status": rule.get("Status") or "Unknown",
                "prefix": rule.get("Filter", {}).get("Prefix"),
            }
            if expiration:
                item["expiration"] = {
                    "days": expiration.get("Days"),
                    "date": isoformat_z(expiration.get("Date")),
                    "expiredObjectDeleteMarker": expiration.get("ExpiredObjectDeleteMarker"),
                }
            if transitions:
                item["transitions"] = [
                    {
                        "days": t.get("Days"),
                        "date": isoformat_z(t.get("Date")),
                        "storageClass": t.get("StorageClass") or "Unknown",
                    }
                    for t in transitions
                ]
            if noncurrent:
                item["noncurrentVersionExpiration"] = {
                    "days": noncurrent.get("NoncurrentDays"),
                    "newerNoncurrentVersions": noncurrent.get("NewerNoncurrentVersions"),
                }
            if abort_upload:
                item["abortIncompleteMultipartUpload"] = {
                    "daysAfterInitiation": abort_upload.get("DaysAfterInitiation")
                }
            rules_out.append(item)
        result["lifecycleRules"] = rules_out
    except Exception as error:
        if not _is_lifecycle_not_configured(error):
            print(f"Failed to get bucket lifecycle: {error}")
            result["lifecycleError"] = str(error)
    return result
